vendors/espressif/boards/esp32/ports/ble/nimble/iot_ble_hal_gatt_server.c

```diff
@@ -727,69 +766,41 @@ static uint16_t prvCountDescriptor( BTService_t * pxService,
  *
  * Since all memory is initialized to 0, that function can be called to clean even half created service.
  */
-static void prvCleanupService( BTService_t * pxService,
-                               struct ble_gatt_svc_def * pSvc )
+static void prvCleanupService( struct ble_gatt_svc_def * pSvc )
 {
-    uint16_t charCount;
-    uint16_t dscrCount;
-    struct ble_gatt_dsc_def * pDescriptors;
-    uint16_t index;
-
-    charCount = 0;
-    dscrCount = 0;
+    const struct ble_gatt_dsc_def * pDescriptor;
+    const struct ble_gatt_chr_def * pCharacteristic;
+    size_t cIndex, dIndex;
 
     if( pSvc->uuid != NULL )
     {
         vPortFree( ( void * ) pSvc->uuid );
     }
 
-    for( index = 0; index < pxService->xNumberOfAttributes; index++ )
+    if( pSvc->characteristics != NULL )
     {
-        switch( pxService->pxBLEAttributes[ index ].xAttributeType )
+        for( cIndex = 0; pSvc->characteristics[ cIndex ].uuid != NULL; cIndex++ )
         {
-            case eBTDbCharacteristic:
-
-                /* Allocate memory for UUID and copies it. */
-                if( pSvc->characteristics[ charCount ].uuid != NULL )
-                {
-                    vPortFree( ( void * ) pSvc->characteristics[ charCount ].uuid );
-                }
-
-                /* If last characteristic had a descriptor array, we can now remove it. */
-                if( dscrCount != 0 )
-                {
-                    if( pSvc->characteristics[ charCount - 1 ].descriptors != NULL )
-                    {
-                        vPortFree( ( void * ) pSvc->characteristics[ charCount - 1 ].descriptors );
-                    }
-                }
-
-                charCount++;
-                dscrCount = 0;
-                break;
-
-            case eBTDbDescriptor:
-                pDescriptors = &pSvc->characteristics[ charCount - 1 ].descriptors[ dscrCount ];
+            pCharacteristic = &pSvc->characteristics[ cIndex ];
+            vPortFree( ( void * ) pCharacteristic->uuid );
 
-                if( pDescriptors->uuid != NULL )
+            if( pCharacteristic->descriptors != NULL )
+            {
+                for( dIndex = 0; pCharacteristic->descriptors[ dIndex ].uuid != NULL; dIndex++ )
                 {
-                    vPortFree( ( void * ) pDescriptors->uuid );
+                    pDescriptor = &pCharacteristic->descriptors[ dIndex ];
+                    vPortFree( ( void * ) pDescriptor->uuid );
                 }
 
-                dscrCount++;
-                break;
-
-            default:
-                break;
+                vPortFree( ( void * ) pCharacteristic->descriptors );
+            }
         }
-    }
 
-    if( pSvc->characteristics != NULL )
```

Comment:
Do we really not need to check this any more? What if the malloc failed halfway though the creation, do we leak in that case?

Reply:
The check is now at line 780 before iterating through the list of characteristics and descriptors.